api/src/wfl/source.clj

```diff
@@ -311,7 +300,7 @@
   "Get first unconsumed snapshot from `source` queue."
   [source]
   (when-let [{:keys [snapshot_id] :as _record} (peek-tdr-source-details source)]
-    (datarepo/snapshot snapshot_id)))
+    ["snapshot" (datarepo/snapshot snapshot_id)]))
```

Comment:
`"snapshot"` is a `:type`, right?

Reply:
maybe? Not sure what you mean by a `:type`.

Reply:
I mean a method dispatch value, but not sure which is why I ask.

Reply:
Ah no this isn't for type dispatching. I've changed the type of the objects stored in the source and executor queues to be the pair `[description value]`. I suppose someone could implement a consumer of this to dispatch off the description if they so wished.

This was really what I was trying to get reviewed - how should information be passed down the processing stages? I've decided to use tuples but I could have `assoc`'ed a description into the object being peeked.

Reply:
A pair is probably OK, but it is not clear to me what it means.
(I noticed in other places that `"snapshot"` was a `description`.)
I've used namespaced keywords for stuff like this.
Stages in the data pipeline associate a private key to maps they process.
Other stages act on that key as necessary.
Namespacing prevents clashes among users of the data.
(This is part of what I was trying to explain when discussing protocol/record trade-offs.)

Reply:
Consider `::Headers` and `::Properties` in PTC's JMS code, for example.

Reply:
So I haven't completely made my mind up about what the thing should be. I'm not sure how to convey this without a type system so here goes:

For source->executor, my thinking was that it was kind of like a data constructor for sum types, ie
```haskell
data ThingThatCanBeImportedIntoAWorkspace = Snapshot s
                                                                             |  Entity e
```
This works well because we can define up front all the things that we can import into a workspace. In clojure, I can implement this as the pair `[name value]` then do a case analysis on the name (you're right - it should be a case analysis in `executor.cjl`).

For the executor->sink interface this doesn't work because the types of the outputs are dynamic. I need to traverse the outputs though in order to ingest files into TDR so i need a representation of this somehow. Right now, I'm just using the raw womtool WDL description so the executor becomes a queue of `[Description Workflow]` pairs. I didn't think making `Description` a member of `Workflow` because it feels more like metadata.

Reply:
Also used qualified keywords as you suggested